pageserver/src/tenant_threads.rs

```diff
@@ -1,77 +1,173 @@
 
-        std::thread::sleep(compaction_period);
-        trace!("compaction thread for tenant {} waking up", tenantid);
+        tokio::time::sleep(compaction_period).await;
+        trace!("compaction loop for tenant {} waking up", tenantid);
 
         // Compact timelines
         let repo = tenant_mgr::get_repository_for_tenant(tenantid)?;
-        repo.compaction_iteration()?;
+        let compaction_result =
+            tokio::task::spawn_blocking(move || repo.compaction_iteration()).await;
+        match compaction_result {
+            Ok(Ok(())) => {
+                // success, do nothing
+            }
+            Ok(Err(e)) => {
+                anyhow::bail!(e.context("Compaction failed"));
+            }
+            Err(e) => {
+                anyhow::bail!("Compaction join error {}", e);
+            }
+        }
     }
 
     trace!(
-        "compaction thread stopped for tenant {} state is {:?}",
+        "compaction loop stopped for tenant {} state is {:?}",
         tenantid,
         tenant_mgr::get_tenant_state(tenantid)
     );
     Ok(())
 }
 
+static START_GC_LOOP: OnceCell<Sender<ZTenantId>> = OnceCell::new();
+static START_COMPACTION_LOOP: OnceCell<Sender<ZTenantId>> = OnceCell::new();
+
+/// Spawn a task that will periodically schedule garbage collection until
+/// the tenant becomes inactive. This should be called on tenant
+/// activation.
+pub fn start_gc_loop(tenantid: ZTenantId) -> Result<()> {
+    START_GC_LOOP
+        .get()
+        .unwrap()
+        .blocking_send(tenantid)
+        .unwrap();
+    Ok(())
+}
+
+/// Spawn a task that will periodically schedule compaction until
+/// the tenant becomes inactive. This should be called on tenant
+/// activation.
+pub fn start_compaction_loop(tenantid: ZTenantId) -> Result<()> {
+    START_COMPACTION_LOOP
+        .get()
+        .unwrap()
+        .blocking_send(tenantid)
+        .unwrap();
+    Ok(())
+}
+
+/// Spawn the TenantTaskManager
+/// This needs to be called before start_gc_loop or start_compaction_loop
+pub fn init_tenant_task_pool() -> Result<()> {
+    let runtime = tokio::runtime::Builder::new_multi_thread()
+        .thread_name("tenant-task-worker")
+        .worker_threads(40) // Way more than necessary
+        .max_blocking_threads(100) // Way more than necessary
+        .enable_all()
+        .build()?;
+
+    let (gc_send, mut gc_recv) = mpsc::channel::<ZTenantId>(100);
+    START_GC_LOOP.set(gc_send).unwrap();
+
+    let (compaction_send, mut compaction_recv) = mpsc::channel::<ZTenantId>(100);
+    START_COMPACTION_LOOP.set(compaction_send).unwrap();
+
+    thread_mgr::spawn(
+        ThreadKind::TenantTaskManager,
+        None,
+        None,
+        "WAL receiver manager main thread",
+        true,
+        move || {
+            runtime.block_on(async move {
+                loop {
+                    tokio::select! {
+                        _ = thread_mgr::shutdown_watcher() => break,
```

Comment:
> break

That would never stop either of the `select!` ed tasks: we drop their handles, returned from `tokio::spawn` and since don't pass any other cancellation channels to those loops, loose any cancellation/await control on those tasks.

Neither we are now able to determine if any of those tasks had finished abruptly with some panic before (and restart those), due to similar reasons.

I think your neighbour comment is related to this, but it's safer to notice.

Reply:
To elaborate on the cancellations more: we cannot do `handle.abort` on the blocking task, but we can
* periodically `(&mut handle).await` the task to see if it's finished or not
* pass the `tokio::sync::watch::Receiver<()>`/`oneshot`/whatnot channels into the threads and to the cancellation manually there

Reply:
> we cannot do handle.abort on the blocking task

cant we? It should be possible, but this would be a noop, so await on a handle will wait for task completion anyway

> periodically (&mut handle).await

I would suggest using `handle.is_finished`. 

> into the threads and to the cancellation manually there

Maybe we can integrate this with our thread_mgr and use shutdown checks from there